### ISSUE TITLE
the Merkleization section of simple-serialize.md

### ISSUE THREAD

**Issue by booleanfunction** [ethereum/eth2.0-specs#1893]
In [simple-serialize.md](https://github.com/ethereum/eth2.0-specs/blob/dev/ssz/simple-serialize.md), within the Merkelization section, a `chunk_count` function is defined as:

>- chunk_count(type): calculate the amount of leafs for merkleization of the type.
>   - all basic types: 1
>   - Bitlist[N] and Bitvector[N]: (N + 255) // 256 (dividing by chunk size, rounding up)
>   - List[B, N] and Vector[B, N], where B is a basic type: (N * size_of(B) + 31) // 32 (dividing by chunk size, rounding up)
>   - List[C, N] and Vector[C, N], where C is a composite type: N
>   - containers: len(fields)

If `chunk_count` does represent _the amount of leafs for merkleization of a type_ then the formulas for `Bitlist[N]`, `List[B,N]` and `List[C,N]` should be amended because any type, including say a `List[B,0]`, requires at least 1 leaf (even if that 1 leaf effectively becomes the root node when calculating the root tree hash).

Note that empty bitvectors and vectors are illegal and so their `chunk_count` will always be at least 1 anyway. Hence the fix can be easily applied using the current grouping of types to specify the various formulas.

**FIX 1: Amend formulas to return at minimum of 1 chunk.**

py-ssz: I can see that in the end everything works out in py-ssz, despite with this mismatch of definition and formula within the specification of `chunk_count`, but it would be great if the spec could be made as precise as possible.

## Further issues

There are several problems (including a missing case and error preventing merkleization of bitlists[N=0] and lists[B,N=0]) in the remainder of the merkleization section, basically stemming from this `chunk_count` issue. Though, again, I note that while everything works out in the py-ssz implementation, from a specification perspective it would be great to fix these problems; therefore ensuring that the spec is precise and consistent.

I will summarise these inconsistencies and then provide a suggested amendment at the end.

## 1.	pack(value) and pack_bits(bits) functions

> - pack(value): given ordered objects of the same basic type, serialize them, pack them into BYTES_PER_CHUNK-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.
> - pack_bits(bits): Given the bits of bitlist or bitvector, get bitfield_bytes by packing them in bytes and aligning to the start. The length-delimiting bit for bitlists is excluded. And then pack bitfield_bytes into BYTES_PER_CHUNK-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.

- For `List[B, N=0]` and `Bitlist[N=0]` the current wording of the spec gives `chunk_count=0`for these cases.
- The wording of the `pack(value)` and `pack_bits(bits)` functions imply that at least 1 chunk will be returned i.e. there must be at least 1 chunk if the _last chunk_ is to be _right-pad with zero bytes_. 
- Hence the current spec implies that `len(pack(value)) > chunk_count(type`) when `value` is a List[B,N=0], and also `len(pack_bits(bits)) > chunk_count(type)` when `bits` represents a Bitlist[N=0].
- This seems counter-intuitive and will cause an error with merkleization as will be discussed below. 
- It makes sense to ensure that padding to whole chunks (such that at least 1 chunk is returned) is performed at the `pack`/`pack_bits` stage.
- Hence, explicitly state that `pack` and `pack_bits` return at least 1 chunk to make the spec precise.
- Also, these points are cited as being further evidence that the `chunk_count` function formulas should be amended so that at least 1 chunk is always returned (see above, Fix 1).
- Note, in py-ssz the `pack` function returns 1 zeroed chunk (the empty chunk) for a list with N=0 but the `pack_bits` function returns [] for a bitlist with N=0. In the bitlist case the padding to 1 empty chunk is left until later, occurring within merkleization. I mention this to acknowledge that the correct output for hash_tree_root is ultimately achieved for these cases in py-ssz but even if implementations differ, the spec should be precise and consistent in the definition of functions.

**FIX 2: Explicitly state that the pack and pack_bits functions return at least 1 chunk.**

## 2.	merkleize(chunks, limit=None) function

> - merkleize(chunks, limit=None): Given ordered BYTES_PER_CHUNK-byte chunks, merkleize the chunks, and return the root:
>   - The merkleization depends on the effective input, which can be padded/limited:
>       - if no limit: pad the chunks with zeroed chunks to next_pow_of_two(len(chunks)) (virtually for memory efficiency).
>       - if limit > len(chunks), pad the chunks with zeroed chunks to next_pow_of_two(limit) (virtually for memory efficiency).
>        - if limit < len(chunks): do not merkleize, input exceeds limit. Raise an error instead.
>   - Then, merkleize the chunks (empty input is padded to 1 zero chunk):
>      - If 1 chunk: the root is the chunk itself.
>      - If > 1 chunks: merkleize as binary tree.

_Case `limit==len(chunks)` not considered_

- The case where `limit == len(chunks)` is not considered in the first step of the merkleization function, yet it is possible for the `limit==len(chunks)` case to occur.  
- Later in the spec we see that limit is set to `chunk_count(type)` for bitvectors, lists of basic objects, bitlists, and lists of composite objects.
- It can also be seen that `len(chunks)=len(pack/pack_bits(value))` for bitvectors, lists of basic objects and bitlists; and `len(chunks)=len([hash_tree_root(element) for element in value])` for lists of composite objects.

- For bitvectors, lists of basic objects and bitlists effectively we are considering the case where `chunk_count(type)==len(pack/pack_bits(value))`:
  - for bitvectors, `chunk_count(type)==len(pack_bits(value))` will always be true; and,
  - for bitlists and lists of basic objects `chunk_count(type)` represents the maximum number of chunks required given the parameter `N`, and there is no reason why the number of values in a bitlist or list can’t be equal to `N` thus ensuring that it is possible for `chunk_count(type)==len(pack/pack_bits(value))`.
- For lists of composite objects effectively we are considering the case where `chunk_count(type)==len([hash_tree_root(element) for element in value])` and again there is no reason why a List[C,N] can't have the full N objects that would cause this case to occur.

**FIX 3: Amend `limit > len(chunks)` to be `limit >= len(chunks)`**

_Error implied for bitlists and list when N=0_

- As the spec is currently written an error can occur preventing merkleization of bitlists and lists (of basic objects) when N=0.
- For example, a `bitlist[N=0]` has `chunk_count=0` and if `pack_bits` does return at least 1 chunk (as implied by the wording in the spec) then an error will be raised in the `merkleize` function because `limit < len(chunks)`. 
   - i.e. since in this case `limit=chunk_count(type)=0` and `len(chunks)=len(pack_bits(bits))=1` as implied by the wording of `pack_bits`.
- In py-ssz the `pack_bits` function returns 0 chunks for a bitlist[0], hence `len(chunks)=len(pack_bits(bits))=0` and so no error is raised. Also note that in py-ssz the `merkleize_with_cache` function specifically deals with the case where `limit==0` and returns `ZERO_HASHES[0]`.
- Hence the _error_ identified here is not applicable in py-ssz however, a change to ensure `chunk_count` > 0 would improve consistency and prevent this error within the spec.

**FIX 4: Refer to Fix 1.**

_Empty input shouldn't occur_

- The second step in the merkleization function says ‘Then, merkleize the chunks (empty input is padded to 1 zero chunk)’
- It is not possible to have empty input at this step (assuming Fix 3). 
- The padding with zeroed chunks to the `next_pow_of_two` chunks that occurs in the first step of merkleize ensures at least 1 chunk by the second step because `next_pow_of_two(i)` > 0.
   - i.e. This would be the case even if 0 chunks were received as input because `next_pow_of_two(0)` should equal 1, assuming that Fix 3 is included to ensure that the first step of the merkleization function is always performed. 
- If Fix 3 weren't adopted, for the case when `limit==len(chunks)==0` we could say that if empty input is padded to 1 zeroed chunk during the second step, then the first step doesn't need to be performed.    
  - However, the inclusion of this empty input provision at the second step of the merkleization function doesn't fix the `limit==len(chunks)` case missing in the first step when we have a variable length type holding a maximum N values. 
  - Hence Fix 3 is necessary.

**FIX 5: Remove reference to ‘empty input is padded to 1 zero chunk’**

## 3.	hash_tree_root(value) function

>We now define Merkleization hash_tree_root(value) of an object value recursively:
> - merkleize(pack(value)) if value is a basic object or a vector of basic objects.
> - merkleize(pack_bits(value), limit=chunk_count(type)) if value is a bitvector.
> - mix_in_length(merkleize(pack(value), limit=chunk_count(type)), len(value)) if value is a list of basic objects.
> - mix_in_length(merkleize(pack_bits(value), limit=chunk_count(type)), len(value)) if value is a bitlist.
> - merkleize([hash_tree_root(element) for element in value]) if value is a vector of composite objects or a container.
> - mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value)) if value is a list of composite objects.
> - mix_in_type(merkleize(value.value), value.type_index) if value is of union type.

- The `limit` parameter not required for a bitvector
- As `chunk_count(bitvector)==len(pack_bits(bits)` (i.e `bits` represent a bitvector) it is not necessary to calculate a `limit`, rather the merkelize function can be called with `limit=None`.

**FIX 6: Amend merkleize for bitvectors to `merkleize(pack_bits(value))`**

# Suggested wording for the merkleization section.

The above suggestions would result in the following amended version of this section of the SSZ specification.

## Merkleization
We first define helper functions:

- `size_of(B)`, where `B` is a basic type: the length, in bytes, of the serialized form of the basic type.
- `chunk_count(type)`: calculate the amount of leafs for merkleization of the type.
   - all basic types: `1`
   - `Bitlist[N]` and `Bitvector[N]`: `max(1,(N + 255) // 256)`
   - `List[B, N]` and `Vector[B, N]`, where `B` is a basic type: `max(1,(N * size_of(B) + 31) // 32)` 
   - `List[C, N]` and `Vector[C, N]`, where `C` is a composite type: `max(1,N)`
   - containers: `len(fields)`
- `pack(value)`: given ordered objects of the same basic type, serialize them, pack them into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks. At least 1 chunk is returned.
- `pack_bits(bits)`: Given the `bits` of bitlist or bitvector, get `bitfield_bytes` by packing them in bytes and aligning to the start. The length-delimiting bit for bitlists is excluded. And then pack `bitfield_bytes` into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks. At least 1 chunk is returned.
- `next_pow_of_two(i)`: get the next power of 2 of `i`, if not already a power of 2, with 0 mapping to 1. Examples: `0->1, 1->1, 2->2, 3->4, 4->4, 6->8, 9->16`
- `merkleize(chunks, limit=None)`: Given ordered `BYTES_PER_CHUNK`-byte chunks, merkleize the chunks, and return the root:
   - The merkleization depends on the effective input, which can be padded/limited:
      - if no limit: pad the `chunks` with zeroed chunks to `next_pow_of_two(len(chunks))` (virtually for memory efficiency).
      - if `limit >= len(chunks)`, pad the chunks with zeroed chunks to `next_pow_of_two(limit)` (virtually for memory efficiency).
      - if `limit < len(chunks)`: do not merkleize, input exceeds limit. Raise an error instead.
   - Then, merkleize the chunks:
      - If `1` chunk: the root is the chunk itself.
      - If `> 1` chunks: merkleize as binary tree.
- `mix_in_length`: Given a Merkle root `root` and a length `length` (`"uint256"` little-endian serialization) return `hash(root + length)`.
- `mix_in_type`: Given a Merkle root `root` and a type_index `type_index` (`"uint256"` little-endian serialization) return `hash(root + type_index)`.

We now define Merkleization `hash_tree_root(value)` of an object `value` recursively:
- `merkleize(pack(value))` if `value` is a basic object or a vector of basic objects.
- `merkleize(pack_bits(value))` if `value` is a bitvector.
- `mix_in_length(merkleize(pack(value), limit=chunk_count(type)), len(value))` if `value` is a list of basic objects.
- `mix_in_length(merkleize(pack_bits(value), limit=chunk_count(type)), len(value))` if `value` is a bitlist.
- `merkleize([hash_tree_root(element) for element in value])` if `value` is a vector of composite objects or a container.
- `mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))` if `value` is a list of composite objects.
- `mix_in_type(merkleize(value.value), value.type_index)` if `value` is of union type.

**Comment by protolambda** [comment]
Hey, thank you for looking deeper into this, here are my thoughts:

> If chunk_count does represent the amount of leafs for merkleization of a type then the formulas for Bitlist[N], List[B,N] and List[C,N] should be amended because any type, including say a List[B,0], requires at least 1 leaf

Having 1 leaf is not the same as having 0. 1 leaf would mean a single *usable* node in as the root of the content-tree (i.e. without the mix-in). Whereas a 0 count just means that there will only be padding to go to the next power of 2 (= 1 here). So you get an unusable, always-zero, node as the root of the content-tree. So it should definitely have a 0 minimum count to reflect that. 

> Note that empty bitvectors and vectors are illegal and so their chunk_count will always be at least 1 anyway. Hence the fix can be easily applied using the current grouping of types to specify the various formulas.

I don't think the fix is necessary, but it can definitely be worded better to avoid confusion.

>  Explicitly state that the pack and pack_bits functions return at least 1 chunk.

Yes, makes sense. But instead of making the chunk-count return something weird (0 data having 1 chunk), it should just fallback to a zeroed chunk at the very last step when there is no data at all.

> CASE limit==len(chunks) NOT CONSIDERED

Yes, no need to scream in capitals, it's the easiest case: when it fully fills the subtree, it can be used as-is, and nothing needs to change to the input. "which *can* be padded/limited: ", in this case, it is not padded or limited. The spec should be more clear about this, but screaming is not the way to change it.

>  Amend limit > len(chunks) to be limit >= len(chunks)

The padding would be a no-op, so I dislike this change. Stating that no padding or limit applied when `limit == len(chunks)` would be enough.

> For example, a bitlist[N=0] has chunk_count=0 and if pack_bits does return at least 1 chunk (as implied by the wording in the spec) then an error will be raised in the merkleize function because limit < len(chunks)

You just introduced this bug if you make chunk count non 0.

> EMPTY INPUT SHOULDN’T OCCUR

Instead of screaming at something which is only broken in so far other non-tested fixes are applied, we could work out the edge cases of each function and decide where to clarify or emphasize things in the spec.

> The limit parameter not required for a bitvector ....

The whole spec is trying to navigate a fine line between minimalism and consistency. Limiting everything to a single merkleization function caused this unrequired parameter. But I don't think it's a bug, nor does omitting it clarify much. Instead, maybe we should separate the merkleization description more, to be less densely packed, and avoid things like this.

> Suggested wording for the merkleization section.

Thank you for suggesting changes, but it really helps if you use a diff instead of copying the full thing over. Or submit a PR. I diffed it manually, and it basically boils down to those few "minimum 1 chunk" changes, and omitting that bitvector limit.

---

And some meta-feedback:
Using capitals and repeating things until they are hard to parse does not help. It comes over aggressive even. Outline the problem you see, ask why it is like this, and try to be constructive.

Meanwhile we have an open draft for a more elaborate SSZ spec, and plan to move SSZ out of this eth2 specs repository. See https://github.com/protolambda/eth2.0-ssz/

This draft is far from perfect yet, but the intention is to improve it, and give SSZ a place where we can go into depth in the more advanced parts of SSZ (partials, multiproofs, backings) and be more welcoming to changes in description, clarification, etc. Given its current state in phase0 in the specs, I'm avoiding making unnecessary changes here.

**Comment by booleanfunction** [comment]
My apologies for the use of capitals, I had no intention to use them in that way, I was just thinking they created a different style of heading and so I used them with the intention of formatting to break up the sections. I am still getting used to using github and so I apologise if this caused any offence, I will think more carefully about how I format in the future :)

**Comment by franck44** [comment]
@protolambda 
thanks for your feedback. I am not sure using capitals mean screaming. It seems to me that the point is to make it clear what problems are uncovered and to structure the exposition of the issue.

Apart from that, I am surprised that we may _think_ something is a bug or not. It is or it is not.
But to decide we need a non-ambiguous reference, for instance a **formal specification**, which the Python-like specs are not.

> This draft is far from perfect yet, but the intention is to improve it, and give SSZ a place where we can go into depth in the more advanced parts of SSZ (partials, multiproofs, backings) and be more welcoming to changes in description, clarification, etc. Given its current state in phase0 in the specs, I'm avoiding making unnecessary changes here.

This is extremely confusing to have so many different specifications. Which one is the most reliable one? Why is the new one better than the other ones? If you write a new one, does it mean that the other ones are not satisfactory?
When does it end?

We are also trying hard to improve and **formally prove** some properties of the specs in [this repo](https://github.com/PegaSysEng/eth2.0-dafny). The objective is to increase confidence in the consistency of the specs by using state-of-the-art formal methods and contribute to the ETH ecosystem.
The issues we have uncovered are non-trivial, and reveal a number of inconsistencies (including in some implementations like py-ssz). 
Our specification is the **ONLY** (no screaming, highlighting) formal version of the specs with clear definitions and mathematical (machine-checkable) **correctness proofs** (the K-specs version does not provide formal proofs, just a semantics of the specs). 
The analysis above (that seems to be dismissed rather quickly) is well articulated, and is the result of extremely careful analysis of the existing specs, and a rigorous (logical) treatment.
This may deserve some attention.

As I mentioned in another issue, the specs would probably benefit from a more formal and thorough treatment rather than another Python-centric (no intended offence here, just a fact) code.
Specs are supposed to be open and not readable only by Python experts. 
For example, do you think the specs can be used to implement a client in Haskell, Scala, Idris, Agda or any functional language? I am not convinced this is the case. They should probably cater for different tastes and inclination in a programming language.
A structured and rigorous approach like the one we are adopting in our formal specs may have some advantages over sometimes hand-wavy and imprecise explanations.

**Comment by protolambda** [comment]
> I am not sure using capitals mean screaming

I know you all intend well, but using full capitals for something is generally not how someone likes to read a conversation, and is often used for screaming in other contexts.

> we may think something is a bug or not. It is or it is not.

The specification may not be formal enough, and I fully agree we should improve it, but if the intention is the same, it is more of a spec issue than a bug.

>  Which one is the most reliable one? 

The one right here in the specs. The other one is a draft.

> Why is the new one better than the other ones? 

Because this one is very dense, misses a lot of insight and information in futures that are part of SSZ, but not in phase0, and is taking formal verification to get right.

> If you write a new one, does it mean that the other ones are not satisfactory?

The current one in the spec has been sufficient for now, but when it comes to describing the ideas, principles and advanced features of SSZ, it is incomplete.

> When does it end?

So far the only users are client implementers, who have tests and are using a stable spec for more than half a year. The extension/improvements of the SSZ spec have not been a priority because of this, as there are other more pressing parts in both the specs as the implementation. That said, now that you are giving it a nudge, I am working to update the draft and get it shipped.

> that seems to be dismissed rather quickly

Sorry, I have not closed the issue, and do not dismiss it. What I am trying to say is that the proposed solution is not improving the spec beyond fitting your interpretation in formally correct way. What I like to achieve is to improve it to a point where it is clear to someone without digging through it formally.

> As I mentioned in another issue, the specs would probably benefit from a more formal and thorough treatment

I agree and encourage the effort, but with the throughput of issues and implementation details in the past, producing tests, and keeping it readable for developers, the python code had a lot of value.

> For example, do you think the specs can be used to implement a client in Haskell, Scala, Idris, Agda or any functional language? I am not convinced this is the case.

I like Scala and Haskell, but the majority of developers are looking at implemention in other languages. We still try to ignore as many python-specific features. Things like `sum` are clear and readable. But things like a metaclass, or even a regular class with methods, do not have a place in the spec for this reason.

And regarding the related `assert` issue, I am fan of separate verification conditions first as well. However, it has been avoided to not have throw verbose exceptions everywhere (which would only make the spec more like an implementation), and keeping the benefits of an  `assert` during debugging (lots of extra information is captured, it's not just a boolean).

**Comment by franck44** [comment]
> The specification may not be formal enough, and I fully agree we should improve it, but if the intention is the same, it is more of a spec issue than a bug.

I must admit that it is hard for me to see what the difference is between a spec issue and a bug.
The system can crash because of a _spec issue_, correct? 
The specs contain bugs (I can bet some ETH on it), and it is useful to try and identify them early.
Some bugs are even related to type-checking (and we reported a few) and that should be fixed as it is ... a bug.
This current issue raises some concerns about genuine inconsistencies in the Merkleise specs.
These problems do show up when we analyse the specs with rigorous techniques.

> I like Scala and Haskell, but the majority of developers are looking at implemention in other languages. 

What IOHK is doing is probably a counter-example to this claim.
And I would not venture into guessing what the majority of developers do, and think that the majority is right. There were a few critical issues in the past on some Ethereum chains, so we may be more humble regarding the way we approach things.
Again, a spec should be open and usable by any developer irrespective of the programming language of choice. This is actually the main feature of a spec: it tells you **what** has to be computed/achieved but gives the free choice of **how** to compute it.

> We still try to ignore as many python-specific features. Things like sum are clear and readable. But things like a metaclass, or even a regular class with methods, do not have a place in the spec for this reason.

You may have look at this part of the specs in `on_block`:
```
# Make a copy of the state to avoid mutability issues
    assert block.parent_root in store.block_states
    pre_state = store.block_states[block.parent_root].copy()
```

If this `copy()` is not Python-centric I am rather confused. 
The specs as they are now written mix a lot of Python-idioms and in my opinion this is detrimental to clarity.
For instance, the potential mutability issue in the comment is not supported by any concrete justification. Just a sort of an arbitrary choice because the author something may cause a problem. 

The specs contain a lot of clever ideas mixed with low-level implementation details that, in my opinion, act as noise.
Moreover, they often do not define **what** has to be computed but provide an algorithm to compute a result. This is very different to a specification.
Again an SSZ example is [the encoding of Bitlists](https://github.com/ethereum/eth2.0-specs/blob/dev/ssz/simple-serialize.md#bitlistn)
```
array = [0] * ((len(value) // 8) + 1)
for i in range(len(value)):
    array[i // 8] |= value[i] << (i % 8)
array[len(value) // 8] |= 1 << (len(value) % 8)
return bytes(array)
```

Apart from providing a recipe to compute something (based off the `bytes` Python function) this does not provide much insight.
[Here](https://github.com/PegaSysEng/eth2.0-dafny/blob/546c2d9974aff7cd7c99aa2772c6ff95f09d5b4f/src/dafny/ssz/BitListSeDes.dfy#L62) we have provided a functional definition i.e. **what** should be computed rather than how.
This is probably more of a spec than the algorithm above.
And this is not at odds with software development best practices: some comments to explain what is happening, a recursive definition to explain **what** the encoding should be. 

The simple fact that every other month a new SSZ spec pops up (more precise, more refined, but always Python-centric) is a probably a sign that Python is not the only avenue for writing a spec.
Again in my opinion, this is a questionable choice, I have never seen Python used to write specifications. 

Overall, I think diversity (and the work we are doing writing a formal spec in Dafny is part of it) will benefit the specs. 
Our intent is to make it work and to provide a high-level of quality and reliability. The Eth2.0 is a critical system and it is important to get it right.
The comments and proposals at the beginning of this issue are justified and identify real problems and propose validated fixes.

**Comment by protolambda** [comment]
> If this copy() is not Python-centric I am rather confused.

Good point, copying is some kind of edge-case, since there is no clear way of doing it otherwise. We would have to use `deepcopy` (very inefficient, but not important), or specify a `copy(v)->v'` helper function somewhere. This is not the kind of method call I'm thinking about when separating python from specs though. It's readable and simple enough. Defining something more eth2-specific as a method call has much bigger (negative) impact on the spec language subset. If you think we should have a copy helper function instead of this call then we can do that, but this is the first time I have heard `.copy()` as a bad example, so I would rather leave it for what it is.

> Again an SSZ example is the encoding of Bitlists

Yes, I dislike that too, it is very implementation-ish, whereas it could be described in a readable but clear way. However, description comes with a burden of wording every ambiguity, whereas implementation always does one thing. In this sense we may have been lazy as spec writers, but at least client implementers could work with it.
See the new SSZ repo draft for a more description-focused spec, and I welcome any fix or note you can add to that.

> The simple fact that every other month a new SSZ spec pops up

This is not true. There have been a few minor issues that have been resolved, but SSZ has barely changed for almost a year (with the exception of removing functionality, such as signing-root).

What I am trying to achieve however, is to slowly push that SSZ draft spec forward, while working on many other things, and keeping the current spec usable for implementers (which are long past implementation of SSZ, and mostly focused on latest v0.12 changes and testnets). And then ship the new SSZ spec for future use, hopefully resolving many of these ugly implementation-like issues.

> Overall, I think diversity ... will benefit the specs.

Really, you don't have to defend it for me, I am supportive of your effort, but I'd like the fruits of the effort to benefit the readability and formal correctness, instead of twiddling something somewhere (even if it actually fixes a bug), where we need to reference the issue as implementer to understand the change fully. Especially with SSZ, we reached a point where the current spec is too dense, and we could do lots of changes without an implementer fully understanding them. I hope you understand the desire for a new SSZ spec, and we get those bugfixes merged with their full detail :+1:

**Comment by franck44** [comment]
> What I am trying to achieve however, is to slowly push that SSZ draft spec forward, while working on many other things, and keeping the current spec usable for implementers (which are long past implementation of SSZ, and mostly focused on latest v0.12 changes and testnets). And then ship the new SSZ spec for future use, hopefully resolving many of these ugly implementation-like issues.

Our verification effort goes beyond all the testing work that has been going on related to the specs.
It is a several-month 3 person effort that is building on state-of-the-art techniques and tools.
It is the only formal, documented, readable, proved and machine-checkable specification.

>  I hope you understand the desire for a new SSZ spec,

I'll be honest: no I don't understand. 
We may not need yet a new (informal, ambiguous) one. Just one that is formal and readable, and ours is a good candidate, yet it seems to be completely ignored.

**Comment by protolambda** [comment]
>  no I don't understand.

@franck44 See my comment in the other thread: https://github.com/ethereum/eth2.0-specs/issues/1901#issuecomment-645354568

**Comment by booleanfunction** [comment]
@protolambda thank you for the detailed feedback. I wanted to make sure that I gave myself a bit of time to carefully think about it but if it is ok I think it would really help me to clarify a few things. Originally I wasn't sure how to break up some of my queries but hopefully they will make sense :)

Can I please check, in the first step of the merkleization function:

>merkleize(chunks, limit=None): Given ordered BYTES_PER_CHUNK-byte chunks, merkleize the chunks, and return the root:
>- The merkleization depends on the effective input, which can be padded/limited:
>   - if no limit: pad the chunks with zeroed chunks to next_pow_of_two(len(chunks)) (virtually for memory efficiency).
>   - if limit > len(chunks), pad the chunks with zeroed chunks to next_pow_of_two(limit) (virtually for memory efficiency).
>   -  if limit < len(chunks): do not merkleize, input exceeds limit. Raise an error instead.

1. Is the idea that by the end of it we should have a number of chunks that is equal to a power of 2? Or is it still ok not to have a power of 2 chunks at this point and instead correct for this when creating the binary tree, if needed?

The context for my question is that I think I may have misunderstood the intention of this step of the merkelization function. ie. I assumed that there would be a power of 2 chunks by the end of this first step. 

You mentioned in a reply above, in talking about the `limit==len(chunks)` case:

>it's the easiest case: when it fully fills the subtree, it can be used as-is, and nothing needs to change to the input. "which can be padded/limited: ", in this case, it is not padded or limited. 

I was thinking about say, a vector[uint8,96]. When merkleization is called, `limit` would equal its `chunk_count`, which would be 3. Also, as the pack function would create 3 chunks and so len(chunks) would be 3. And so in this example `limit == len(chunks)`.

2. For this example, is the idea that the first step of the merkleization function would not be implemented, as we have the `limit == len(chunks)` case, but rather just the 3 chunks would be left unchanged and we would go to the second part of the merkleization function? i.e. the part
>Then, merkelize the chunks: 

Thank you for your help :)

**Comment by protolambda** [comment]
@booleanfunction The background thay may be missing here is that subtrees of a power of 2, and the approach to padding, have some implications that need to be understood by client implementers. And with the current minimal spec, the prose does not really cover it. The amount of chunks is something that no (production) implementation should actually pad by adding a whole lot of zeroed chunks at the bottom level.

E.g. if you have a list subtree with 1024 chunks as limit, but 513 in chunk length, then you would have 511 chunks of padding.
For spec purposes this is fine, but I want the spec to at least hint that it was designed to be done differently: you only need the top nodes in the subtrees of the padding. So when you are padding 511 chunks, you really only need 9 nodes in the tree: `1+2+4+8+16+32+64+128+256=511`.

And since these subtrees in the padding are all zeroed, these values can be precomputed efficiently (`Z[i] = H(Z[i-1], Z[i-1]`). This is what the "virtually for memory efficiency" is trying to say.
For a type like the validator registry, with a high limit of `2**40`, this feature is actually really important.


I see now how the `==` part is also misleading: "which *can* be padded/limited: " (emphasis mine): we should indeed add a `==` case for clarity, but what really happens here is that since the `length` and `limit` are the same, nothing much happens if it's a power of two already. But if not, like you pointed out (thanks!) it should be clarified that it would involve padding to the next power of two. (which will always be zero due to the limit restricting the actual values).

And then in the original issue we have the question if `pack` and `pack_bits` should return at least one chunk or not. I think we should keep the output strictly `(len(input bytes) + 31) // 32` chunks. Whatever shape or form the input has, it gets packed into chunks of 32 bytes each, independent of its later purpose or context. So 0 input would be 0 chunks.

When `pack` is actually used to build the bottom of the tree, the (maybe virtual) padding of chunks should be handled as part of the merkleization setup. An empty tree has no data, and I think it's best to regard it as just another case where you would fill the void with zeroes. If `limit == 0`, then the `len(chunks) == limit` applies. And if the `limit` is higher, then we fill automatically with the appropriate amount of zeroes (and with the virtual padding thing, this becomes a matter of simply looking up a precomputed value). 

Also, some minor improvements would be to describe the pad input as `values`, and define `pack_bits` in terms of `pack`.

So what we could redefine `pack(value)` as:
- `pack(values)`: given ordered objects of the same basic type:
  1. Serialize the values into bytes
  2. If not aligned to a multiple of `BYTES_PER_CHUNK` bytes, right-pad with zeroes to the next multiple.
  2. Partition the bytes into `BYTES_PER_CHUNK`-byte chunks
  3. Return the chunks.

And then we can define `pack_bits(bits)` as:
- `pack_bits(bits)`: Given the bits of bitlist or bitvector, get `bitfield_bytes` by packing them in bytes and aligning to the start. The length-delimiting bit for bitlists is excluded. Then return `pack(bitfield_bytes)`

I think it's clear that the input of bytes is serialized as-is in the above, and we avoid some duplication of padding logic.

And then update the merkleization to reflect:
- That it must be limited and padded following the rules (remove the "can")
- Change the `limit > len` case to `>=` like you proposed

What do you think? Would like to submit another PR? Alternatively I can make it too, but have my hands full with Altona testnet and networking things at the moment.

**Comment by booleanfunction** [comment]
@protolambda that sounds great and yes, I would be happy to do the PR (it will give me another chance to go through the process :) I will organise the PR today using your suggestions.

And thank you for linking this to pack/pack_bits as well, I was still thinking about that part too! :)

> And then in the original issue we have the question if pack and pack_bits should return at least one chunk or not. I think we should keep the output strictly (len(input bytes) + 31) // 32 chunks. Whatever shape or form the input has, it gets packed into chunks of 32 bytes each, independent of its later purpose or context. So 0 input would be 0 chunks.

I like this, it keeps things nice and simple. I was thinking about my original dilemma about wording for pack/pack_bits and the number of chunks that they return and well, having more understanding now as to the background/intention in conjunction with these clarifications definitely helps to make it more consistent. I think when I saw that in **py-ssz** one empty chunk is returned from pack for a `List[B, 0]` it previously convinced me that it was _always at least 1_ but from a spec perspective, the consistency that comes from making the output from pack/pack_bits match to the corresponding chunk_count (and so 0 input would be 0 chunks) is great. 

Anyway, thank you also for the extra detail relating to an implementation perspective. I am definitely more what I would term a mathematical programmer, rather than a software developer, but I like thinking about optimisation of algorithms and so any extra knowledge I can gain regarding the implementation is much appreciated. One of the things I think is great about our formal verification project is that while we are aiming to keep the core version as simple as possible (and align with the spec as written), we can also set up implementation/optimised versions of functions and definitively prove they are equivalent - which I think is very cool!

Thank you again

**Comment by protolambda** [comment]
Closing this issue, I think #1934 covers all of this. If anything was missed, let's move that into a new issue. Thanks for making the PR :+1:
